pyomo/contrib/parmest/parmest.py

```diff
@@ -229,12 +232,458 @@ def _experiment_instance_creation_callback(
 
 def SSE(model):
     """
-    Sum of squared error between `experiment_output` model and data values
+    Returns an expression that is used to compute the sum of squared error (`SSE`) objective,
+    assuming Gaussian i.i.d. errors
+
+    Argument:
+        model: annotated Pyomo model
     """
-    expr = sum((y - y_hat) ** 2 for y, y_hat in model.experiment_outputs.items())
+    # Check that experimental outputs exist
+    try:
+        measured_variables = [y_hat.name for y_hat, y in model.experiment_outputs.items()]
+    except:
+        raise RuntimeError(
+            "Experiment model does not have suffix " + '"experiment_outputs".'
+        )
+
+    # sum of squared error between the prediction and observation of the measured variables
+    expr = sum((y - y_hat) ** 2 for y_hat, y in model.experiment_outputs.items())
     return expr
 
 
+def SSE_weighted(model):
+    """
+    Returns an expression that is used to compute the `SSE_weighted` objective,
+    assuming Gaussian i.i.d. errors, with measurement error standard deviation defined in the annotated Pyomo model
+
+    Argument:
+        model: annotated Pyomo model
+    """
+    # Check that experimental outputs exist
+    try:
+        measured_variables = [y_hat.name for y_hat, y in model.experiment_outputs.items()]
+    except:
+        raise RuntimeError(
+            "Experiment model does not have suffix " + '"experiment_outputs".'
+        )
+
+    # check that measurement errors exist
+    try:
+        measured_variables_error = [k.name for k, v in model.measurement_error.items()]
+    except:
+        raise RuntimeError(
+            "Experiment model does not have suffix " + '"measurement_error".'
+        )
+
+    # check if all the values of the measurement error standard deviation have been supplied
+    if all(
+        model.measurement_error[y_hat] is not None for y_hat in model.experiment_outputs
+    ):
+        # calculate the weighted SSE between the prediction and observation of the measured variables
+        expr = (1 / 2) * sum(
+            ((y - y_hat) / model.measurement_error[y_hat]) ** 2
+            for y_hat, y in model.experiment_outputs.items()
+        )
+        return expr
+    else:
+        raise ValueError(
+            'One or more values are missing from `measurement_error`.'
+        )
+
+
+class CovMethodLib(Enum):
+    finite_difference = "finite_difference"
+    kaug = "kaug"
```

Comment:
Upon second throught, I think we should change this to "automatic_differentiation_kaug". The reason is subtle. There is also legacy kaug code that computes the reduced hessian. This change in name also anticipates a future contribution from me for "automatic_differentiation_pynumero". I want to have both versions of AD available for a while until we are confident PyNumero completely agrees with kaug and we fully depreciate kaug (if ever) from the codebase.